mbus-api/src/mbus_nats.rs

```diff
@@ -0,0 +1,337 @@
+    /// connection error after some retries.
+    async fn publish(&mut self, message: &EventMessage) -> BusResult<u64> {
+        let options = BackoffOptions::publish_backoff_options();
+        let mut tries = 0;
+        let mut log_error = true;
+
+        let subject = NatsMessageBus::subject(message);
+        let payload = NatsMessageBus::payload(message)?;
+
+        loop {
+            let publish_request = self.jetstream.publish(subject.clone(), payload.clone());
+            let err = match publish_request.await {
+                Ok(x) => match x.await {
+                    Ok(y) => {
+                        return Ok(y.sequence);
+                    }
+                    Err(error) => error,
+                },
+                Err(error) => error,
+            };
+
+            if log_error {
+                tracing::warn!(
+                    "Error publishing message to jetstream: {:?}. Retrying...",
+                    err
+                );
+                let _stream = self.verify_stream_exists().await;
+                log_error = false;
+            }
+
+            if tries == options.max_retries {
+                return Err(Error::PublishError {
+                    retries: options.max_retries,
+                    payload: format!("{message:?}"),
+                    error: err.to_string(),
+                });
+            }
+            if let Some(error) = err.downcast_ref::<std::io::Error>() {
+                if error.kind() == ErrorKind::TimedOut {
+                    tries += 1;
+                    continue;
+                }
+            }
+            backoff_with_options(&mut tries, &options).await;
+        }
+    }
+
+    /// Subscribes to the stream on messages though a consumer. Returns an ordered stream of
+    /// messages published to the stream.
+    async fn subscribe<T: Serialize + DeserializeOwned>(
+        &mut self,
+    ) -> BusResult<BusSubscription<T>> {
+        tracing::debug!("Subscribing to Nats message bus");
+        let stream = self.get_or_create_stream(None).await?;
+
+        let messages = self.create_consumer_and_get_messages(stream).await?;
+        tracing::debug!("Subscribed to Nats message bus successfully");
+
+        return Ok(BusSubscription {
+            messages,
+            _phantom: PhantomData::default(),
+        });
+    }
+}
+
+/// The messages from bus are deserialised into the type T
+pub struct BusSubscription<T> {
+    pub messages: Messages,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Serialize + DeserializeOwned> BusSubscription<T> {
+    pub async fn next(&mut self) -> Option<T> {
+        loop {
+            if let Some(message) = self.messages.next().await {
+                let message = match message {
+                    Ok(message) => message,
+                    Err(error) => {
+                        tracing::warn!("Error accessing jetstream message: {:?}.", error);
```

Comment:
Consider adding fields on left side, %for display and ? for Debug.
```suggestion
                        tracing::warn!(%error, "Error accessing jetstream message");
```
Also trace messages should not end with period `.`